pkg/eosclient/eosclient.go

```diff
@@ -269,76 +265,99 @@ func (c *Client) executeEOS(ctx context.Context, cmd *exec.Cmd) (string, string,
-	acls, err := c.getACLForPath(ctx, username, path)
+func (c *Client) getVersion(ctx context.Context, rootUID, rootGID string) (eosVersion, error) {
+	cmd := exec.CommandContext(ctx, c.opt.EosBinary, "-r", rootUID, rootGID, "version")
+	stdout, _, err := c.executeEOS(ctx, cmd)
 	if err != nil {
-		return err
+		return "", err
 	}
+	return c.parseVersion(ctx, stdout), nil
+}
 
-	// since EOS Citrine ACLs are is stored with uid, we need to convert username to uid
-	// only for users.
-	if a.Type == acl.TypeUser {
-		a.Qualifier, err = getUID(a.Qualifier)
-		if err != nil {
-			return err
+func (c *Client) parseVersion(ctx context.Context, raw string) eosVersion {
+	var serverVersion string
+	rawLines := strings.Split(raw, "\n")
+	for _, rl := range rawLines {
+		if rl == "" {
+			continue
+		}
+		if strings.HasPrefix(rl, "EOS_SERVER_VERSION") {
+			serverVersion = strings.Split(strings.Split(rl, " ")[0], "=")[1]
+			break
 		}
 	}
-	err = acls.SetEntry(a.Type, a.Qualifier, a.Permissions)
-	if err != nil {
-		return err
+
+	if strings.HasPrefix(serverVersion, "4.") {
+		return versionCitrine
 	}
-	sysACL := acls.Serialize()
+	return versionAquamarine
+}
 
-	// setting of the sys.acl is only possible from root user
-	unixUser, err := c.getUnixUser(rootUser)
+// AddACL adds an new acl to EOS with the given aclType.
+func (c *Client) AddACL(ctx context.Context, uid, gid, rootUID, rootGID, path string, a *acl.Entry) error {
+	version, err := c.getVersion(ctx, rootUID, rootGID)
 	if err != nil {
 		return err
 	}
 
-	cmd := exec.CommandContext(ctx, c.opt.EosBinary, "-r", unixUser.Uid, unixUser.Gid, "attr", "-r", "set", fmt.Sprintf("sys.acl=%s", sysACL), path)
+	var cmd *exec.Cmd
+	if version == versionCitrine {
+		sysACL := a.CitrineSerialize()
+		cmd = exec.CommandContext(ctx, c.opt.EosBinary, "-r", rootUID, rootGID, "acl", "--sys", "--recursive", sysACL, path)
+	} else {
+		acls, err := c.getACLForPath(ctx, uid, gid, path)
+		if err != nil {
+			return err
+		}
+
+		err = acls.SetEntry(a.Type, a.Qualifier, a.Permissions)
+		if err != nil {
+			return err
+		}
+		sysACL := acls.Serialize()
+		cmd = exec.CommandContext(ctx, c.opt.EosBinary, "-r", rootUID, rootGID, "attr", "-r", "set", fmt.Sprintf("sys.acl=%s", sysACL), path)
+	}
+
 	_, _, err = c.executeEOS(ctx, cmd)
 	return err
 
 }
 
 // RemoveACL removes the acl from EOS.
-func (c *Client) RemoveACL(ctx context.Context, username, path string, aclType string, recipient string) error {
-	acls, err := c.getACLForPath(ctx, username, path)
+func (c *Client) RemoveACL(ctx context.Context, uid, gid, rootUID, rootGID, path string, a *acl.Entry) error {
+	version, err := c.getVersion(ctx, rootUID, rootGID)
 	if err != nil {
 		return err
 	}
 
-	// since EOS Citrine ACLs are stored with uid, we need to convert username to uid
-	if aclType == acl.TypeUser {
-		recipient, err = getUID(recipient)
+	var cmd *exec.Cmd
+	if version == versionCitrine {
+		sysACL := a.CitrineSerialize()
```

Comment:
This variable should be of the form: `%s:%s=`, to remove the entry:

```
 acl --user u:1010= /eos/dev
```

Reply:
As discussed, permissions would be an empty string when called from RemoveACL so it would be empty